drivers/dac/dac_mcp4725.c

```diff
@@ -0,0 +1,214 @@
+ * found at https://ww1.microchip.com/downloads/en/DeviceDoc/22039d.pdf
+ */
+
+/* Bits that can't be set when an application wants to write to dac */
+#define MCP4725_VALUE_DONT_SET_BITS			0x07FFF000
+
+/* Defines for field values in MCP4725 DAC register */
+#define MCP4725_WRITE_MODE_Mask				0x7
+#define MCP4725_POWER_DOWN_Mask				0x3
+#define MCP4725_DAC_VAL_Mask                            0xFFF
+
+#define MCP4725_FAST_MODE_POWER_DOWN_Pos		4
+#define MCP4725_FAST_MODE_DAC_UPPER_VAL_Pos		8
+#define MCP4725_FAST_MODE_DAC_UPPER_VAL_Mask		0xF
+#define MCP4725_FAST_MODE_DAC_LOWER_VAL_Mask		0xFF
+
+#define MCP4725_WRITE_DAC_MODE_WRITE_MODE_Pos		5
+#define MCP4725_WRITE_DAC_MODE_POWER_DOWN_Pos		1
+#define MCP4725_WRITE_DAC_MODE_DAC_UPPER_VAL_Pos	4
+#define MCP4725_WRITE_DAC_MODE_DAC_UPPER_VAL_Mask	0xFF
+#define MCP4725_WRITE_DAC_MODE_DAC_LOWER_VAL_Pos	4
+#define MCP4725_WRITE_DAC_MODE_DAC_LOWER_VAL_Mask	0xF0
+
+#define MCP4725_READ_RDY_Pos				7
+#define MCP4725_READ_RDY_Mask				0x1
+
+/* After writing eeprom, the MCP4725 can be in a busy state for 25 - 50ms
+ * See section 1.0 of MCP4725 datasheet, 'Electrical Characteristics'
+ */
+#define MCP4725_BUSY_TIMEOUT_MS                         60
+
+struct mcp4725_config {
+	const char *i2c_master_dev_name;
+	uint16_t i2c_slave_addr;
+};
+
+struct mcp4725_data {
+	const struct device *i2c_master;
+};
+
+/* Read mcp4725 and check RDY status bit */
+static int mcp4725_wait_until_ready(const struct device *dev, uint16_t slave_addr)
+{
+	uint8_t rxData[5];
+	bool mcp4725_ready = false;
+	int32_t retVal;
+	int32_t timeout = k_uptime_get_32() + MCP4725_BUSY_TIMEOUT_MS;
+
+	/* Wait until RDY bit is set or return error if timer exceeds MCP4725_BUSY_TIMEOUT_MS */
+	while (!mcp4725_ready) {
+		retVal = i2c_read(dev, rxData, sizeof(rxData), slave_addr);
+
+		if (retVal == 0) {
+			mcp4725_ready = (rxData[0] >> MCP4725_READ_RDY_Pos) & MCP4725_READ_RDY_Mask;
+		} else {
+			/* I2C error */
+			return retVal;
+		}
+
+		if (k_uptime_get_32() > timeout) {
+			return -ETIMEDOUT;
+		}
+	}
+
+	return 0;
+}
+
+/* MCP4725 is a single channel 12 bit DAC */
+static int mcp4725_channel_setup(const struct device *dev,
+				   const struct dac_channel_cfg *channel_cfg)
+{
+	if (channel_cfg->channel_id != 0) {
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != 12) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int mcp4725_write_value(const struct device *dev, uint8_t channel,
+				uint32_t value)
+{
+	struct mcp4725_data *data = (struct mcp4725_data *)dev->data;
+	const struct mcp4725_config *config = (struct mcp4725_config *)dev->config;
+	uint8_t writeMode;
+	uint8_t powerDownBits;
+	uint16_t dacVal;
+	uint32_t retVal;
+
+	if (channel != 0) {
+		return -EINVAL;
+	}
+
+	/* Check value isn't over 12 bits */
+	if ((value & MCP4725_VALUE_DONT_SET_BITS) != 0) {
+		return -ENOTSUP;
+	}
+	writeMode = (value >> MCP4725_WRITE_MODE_BIT_SHIFT) & MCP4725_WRITE_MODE_Mask;
```

Comment:
The value passed to this function cannot contain anything else than the raw DAC value. If we really need to support both different modes, we need to think about a different solution (e.g. via devicetree).

Reply:
I've removed the configuration bits and with it the "WRITE_TO_EEPROM" mode. Only extended functionality was removed so this dac still functions with the standard interface. I'm open to adding this functionality back in a more suitable way in a later pull request. I may need some hints on what to do there - if it's added to the device tree does that mean the extended configuration is set at compile time? Is there a way to add optional functions to the dac_driver_api interface?